### DISCUSSION TITLE
gctoolkit integration with jar

### DISCUSSION THREAD

**Discussion by AlinaHovakimyan** [microsoft/gctoolkit#351] [Q&A] [answered]
Hi everyone, my mvn project works fine with gctoolkit on intelij, but when I create a jar from it, gctoolkit stops to see my implementations/provides for the Aggregation interface. Does anybody know what could be the issue?
here is my build configuration on maven: 
` <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <version>3.6.1</version>
                <executions>
                    <execution>
                        <id>copy-dependencies</id>
                        <phase>package</phase>
                        <goals>
                            <goal>copy-dependencies</goal>
                        </goals>
                        <configuration>
                            <includeScope>runtime</includeScope>
                            <outputDirectory>${project.build.directory}/libs</outputDirectory>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.3.0</version>
                <configuration>
                    <finalName>GCTK_Integration</finalName>
                    <archive>
                        <manifest>
                            <mainClass>org.example.Main</mainClass>
                            <addClasspath>true</addClasspath>
                            <classpathPrefix>libs</classpathPrefix>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>
        </plugins>
    </build>`
    

And the output I get (there seem some other wornings as well, i just leaved them for the full referance):
`% java -jar .//target/GCTK_Integration.jar 
Hello
Apr 30, 2024 1:01:01 PM io.netty.resolver.dns.DnsServerAddressStreamProviders <clinit>
WARNING: Can not find io.netty.resolver.dns.macos.MacOSDnsServerAddressStreamProvider in the classpath, fallback to system defaults. This may result in incorrect DNS resolutions on MacOS. Check whether you have a dependency on 'io.netty:netty-resolver-dns-native-macos'
Apr 30, 2024 1:01:01 PM com.microsoft.gctoolkit.jvm.AbstractJavaVirtualMachine analyze
INFO: No Aggregations have been registered, DataSource will not be analysed.
Apr 30, 2024 1:01:01 PM com.microsoft.gctoolkit.jvm.AbstractJavaVirtualMachine analyze
INFO: Is there a module containing Aggregation classes on the module-path
Apr 30, 2024 1:01:01 PM com.microsoft.gctoolkit.jvm.AbstractJavaVirtualMachine analyze
INFO: Is GCToolKit::loadAggregationsFromServiceLoader() or GCToolKit::loadAggregation(Aggregation) being invoked?
The provided log file is a G1 GC log`

**Comment by dsgrieve** [comment, marked as answer]
> Apr 30, 2024 1:01:01 PM com.microsoft.gctoolkit.jvm.AbstractJavaVirtualMachine analyze INFO: Is there a module containing Aggregation classes on the module-path 

@AlinaHovakimyan There are three ways that you can get gctoolkit to find your Aggregation classes. 

1. Call the `GCToolKit#loadAggregation(Aggregation)` method
2. Use Java modules for your application and add the `provides` descriptor in a module-info.java file for the Aggregations your application provides. 
3. Add service provider configuration files for your Aggregations to the `META-INF\services` directory of your jar file.

For the service provider approaches, you'll need to call `GCToolKit#loadAggregationsFromServiceLoader()`

There is also a `sample` directory in gctoolkit that has an example module-info.
